# 1- Business Understanding

## Project Overview
Recommendation systems are among the most widely used artificial intelligence applications in modern digital platforms such as Netflix, YouTube, Spotify, and Amazon. Their primary goal is to help users discover relevant content efficiently by analyzing similarities between items or user preferences.

In this project, we aim to build an advanced movie recommendation system capable of suggesting similar movies based on metadata and textual information such as genres, keywords, cast, crew, and movie overviews.

The project combines Natural Language Processing (NLP), feature engineering, vectorization techniques, and similarity learning to simulate real-world AI recommendation systems.

---

## Business Problem
Modern streaming platforms contain massive amounts of content, making it difficult for users to find relevant movies efficiently. Without intelligent recommendation systems, users may experience information overload, lower engagement, and reduced satisfaction.

The objective of this project is to solve this problem by developing a recommendation engine that can intelligently recommend movies similar to a movie selected by the user.

---

## Project Objective
The main objective of this project is to create a content-based movie recommendation system that:

- Understands movie content using NLP techniques
- Extracts meaningful textual and metadata features
- Measures similarity between movies
- Generates personalized movie recommendations
- Simulates real-world recommendation systems used by major technology companies

---

## Recommendation System Type
This project mainly focuses on:

### Content-Based Filtering
The recommendation system suggests movies based on similarities in:
- Genres
- Keywords
- Movie descriptions
- Cast members
- Directors
- Metadata features

The system learns relationships between movies themselves rather than user behavior.

---

## Dataset Description
The project uses the TMDB 5000 Movies Dataset, which contains:
- Movie metadata
- Genres
- Keywords
- Cast and crew information
- Movie overviews
- Popularity metrics
- Ratings and vote counts

The dataset provides rich textual and categorical information suitable for NLP and recommendation system development.

---

## Project Workflow
The project will follow a complete end-to-end recommendation system pipeline:

1. Business Understanding
2. Data Collection and Understanding
3. Data Cleaning and NLP Preprocessing
4. Feature Engineering
5. Text Vectorization
6. Similarity Learning
7. Recommendation Engine Development
8. Evaluation and Recommendation Analysis
9. Advanced Recommendation Improvements
10. Deployment using Streamlit

---

## Technologies and Libraries
The project will mainly use:
- Python
- Pandas
- NumPy
- Scikit-learn
- NLTK
- NLP Vectorizers
- Cosine Similarity
- Streamlit

---

## Expected Outcome
At the end of this project, the system should be able to:
- Recommend movies similar to a selected movie
- Understand textual movie information using NLP
- Compute movie similarity efficiently
- Generate intelligent recommendations
- Be deployed as an interactive web application

---

## Real-World Importance
Recommendation systems are critical components in modern AI products because they improve:
- User experience
- Engagement
- Retention
- Personalization
- Platform revenue

This project demonstrates practical AI engineering concepts commonly used in real-world recommendation systems and modern machine learning applications.

# 2- Imports & Environment Setup :
In this section, we will import all the necessary Python libraries required for data manipulation, visualization, machine learning, and model evaluation, NLP processing and more.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
sns.set_style('whitegrid')

In [2]:
import nltk
from nltk.stem.porter import PorterStemmer
from nltk.corpus import stopwords

In [3]:
import ast

In [4]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

In [5]:
import warnings
warnings.filterwarnings('ignore')

In [75]:
import string

In [167]:
from gensim.models import Word2Vec
from nltk.tokenize import word_tokenize

In [156]:
nltk.download('punkt_tab')

[nltk_data] Downloading package punkt_tab to
[nltk_data]     C:\Users\walid\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping tokenizers\punkt_tab.zip.


True

# 3- Loading the Dataset and understanding it :

In [7]:
# 3.1- Data Loading:
movies = pd.read_csv('../Data/tmdb_5000_movies.csv')
movies.head(2)

,budget,genres,homepage,id,keywords,original_language,original_title,overview,popularity,production_companies,production_countries,release_date,revenue,runtime,spoken_languages,status,tagline,title,vote_average,vote_count
0,237000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.avatarmovie.com/,19995,"[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...",en,Avatar,"In the 22nd century, a paraplegic Marine is di...",150.437577,"[{""name"": ""Ingenious Film Partners"", ""id"": 289...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2009-12-10,2787965087,162.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,Enter the World of Pandora.,Avatar,7.2,11800
1,300000000,"[{""id"": 12, ""name"": ""Adventure""}, {""id"": 14, ""...",http://disney.go.com/disneypictures/pirates/,285,"[{""id"": 270, ""name"": ""ocean""}, {""id"": 726, ""na...",en,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...",139.082615,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}, {""...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2007-05-19,961000000,169.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"At the end of the world, the adventure begins.",Pirates of the Caribbean: At World's End,6.9,4500


In [8]:
# creadit datasets:
credits = pd.read_csv('../Data/tmdb_5000_credits.csv')
credits.head(2)

,movie_id,title,cast,crew
0,19995,Avatar,"[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."
1,285,Pirates of the Caribbean: At World's End,"[{""cast_id"": 4, ""character"": ""Captain Jack Spa...","[{""credit_id"": ""52fe4232c3a36847f800b579"", ""de..."


In [19]:
# seeing the data columns:
movies.columns

Index(['budget', 'genres', 'homepage', 'id', 'keywords', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'vote_average',
       'vote_count'],
      dtype='object')

In [20]:
credits.columns

Index(['movie_id', 'title', 'cast', 'crew'], dtype='object')

In [21]:
# checking the columns data types :
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4803 non-null   int64  
 1   genres                4803 non-null   object 
 2   homepage              1712 non-null   object 
 3   id                    4803 non-null   int64  
 4   keywords              4803 non-null   object 
 5   original_language     4803 non-null   object 
 6   original_title        4803 non-null   object 
 7   overview              4800 non-null   object 
 8   popularity            4803 non-null   float64
 9   production_companies  4803 non-null   object 
 10  production_countries  4803 non-null   object 
 11  release_date          4802 non-null   object 
 12  revenue               4803 non-null   int64  
 13  runtime               4801 non-null   float64
 14  spoken_languages      4803 non-null   object 
 15  status               

In [22]:
credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   movie_id  4803 non-null   int64 
 1   title     4803 non-null   object
 2   cast      4803 non-null   object
 3   crew      4803 non-null   object
dtypes: int64(1), object(3)
memory usage: 150.2+ KB


- we notice that we have jason type data as strings like : genres , production_companies, production_countries, keywords, spoken_languages.
- same thing for creadit : cast and crew.

In [23]:
# checking the shpe of the datasets:
movies.shape

(4803, 20)

In [25]:
credits.shape

(4803, 4)

In [26]:
# statistical sammury :
movies.describe()

,budget,id,popularity,revenue,runtime,vote_average,vote_count
count,4.803000e+03,4803.000000,4803.000000,4.803000e+03,4801.000000,4803.000000,4803.000000
mean,2.904504e+07,57165.484281,21.492301,8.226064e+07,106.875859,6.092172,690.217989
std,4.072239e+07,88694.614033,31.816650,1.628571e+08,22.611935,1.194612,1234.585891
min,0.000000e+00,5.000000,0.000000,0.000000e+00,0.000000,0.000000,0.000000
25%,7.900000e+05,9014.500000,4.668070,0.000000e+00,94.000000,5.600000,54.000000
50%,1.500000e+07,14629.000000,12.921594,1.917000e+07,103.000000,6.200000,235.000000
75%,4.000000e+07,58610.500000,28.313505,9.291719e+07,118.000000,6.800000,737.000000
max,3.800000e+08,459488.000000,875.581305,2.787965e+09,338.000000,10.000000,13752.000000


- we can see that some columns have min 0, which mean information wasn't available(missing value) like in budjet , revenue and runtime ...
- and for jason format we have empty list [].
- we should deal with those later to make the datset complete.

In [34]:
credits[credits['crew'] == '[]'].head(2)

,movie_id,title,cast,crew
3661,19615,Flying By,"[{""cast_id"": 1, ""character"": ""George"", ""credit...",[]
3670,447027,Running Forever,[],[]


- we can see the same thing here in credit dataset, for crew and cast culd be empty list for non available data.

In [35]:
# checking for missing data:
movies.isnull().sum()

budget                     0
genres                     0
homepage                3091
id                         0
keywords                   0
original_language          0
original_title             0
overview                   3
popularity                 0
production_companies       0
production_countries       0
release_date               1
revenue                    0
runtime                    2
spoken_languages           0
status                     0
tagline                  844
title                      0
vote_average               0
vote_count                 0
dtype: int64

In [37]:
credits.isnull().sum()

movie_id    0
title       0
cast        0
crew        0
dtype: int64

- we are goning to deal with missing values later, also the cases we specified for empty list and 0 values.

In [38]:
# checking for duplicted rows:
movies.duplicated().sum()

np.int64(0)

In [39]:
credits.duplicated().sum()

np.int64(0)

- we are good here, we don't have any duplicated rows.

# 4- Cleaning and preparing the final dataset:

# 4.1- Merging the two dataset :

In [40]:
# we merge movies and credit on the title column :
movies = movies.merge(credits, on='title')

# 4.2- selecting the final features for the dat set:

- Because recommendation systems need: high-quality semantic features not noisy metadata, and reudndant and duplicated features:

In [42]:
#we chose the right mainingful features:
movies.columns

Index(['budget', 'genres', 'homepage', 'id', 'keywords', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'vote_average',
       'vote_count', 'movie_id', 'cast', 'crew'],
      dtype='object')

- final set of features will be like this : movie_id -->	unique identifier, title --> used for recommendations/search
overview --> main NLP semantic feature, genres --> movie category similarity, keywords --> theme similarity, cast --> actor similarity
crew --> director similarity

In [43]:
movies = movies[['movie_id',
                 'title',
                 'overview',
                 'genres',
                 'keywords',
                 'cast',
                 'crew']]

In [44]:
movies.head(2)

,movie_id,title,overview,genres,keywords,cast,crew
0,19995,Avatar,"In the 22nd century, a paraplegic Marine is di...","[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...","[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...","[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."
1,285,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...","[{""id"": 12, ""name"": ""Adventure""}, {""id"": 14, ""...","[{""id"": 270, ""name"": ""ocean""}, {""id"": 726, ""na...","[{""cast_id"": 4, ""character"": ""Captain Jack Spa...","[{""credit_id"": ""52fe4232c3a36847f800b579"", ""de..."


# 4.3- Dealing with missing data in this final datset:

In [45]:
movies.isnull().sum()

movie_id    0
title       0
overview    3
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [46]:
# so small amount so droping the safet solution here :
movies.dropna(subset=['overview'], inplace=True)

In [47]:
movies.isnull().sum()

movie_id    0
title       0
overview    0
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [48]:
# reseting the index after droping:
movies.reset_index(drop=True, inplace=True)

In [49]:
# final datset:
movies.head(2)

,movie_id,title,overview,genres,keywords,cast,crew
0,19995,Avatar,"In the 22nd century, a paraplegic Marine is di...","[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...","[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...","[{""cast_id"": 242, ""character"": ""Jake Sully"", ""...","[{""credit_id"": ""52fe48009251416c750aca23"", ""de..."
1,285,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...","[{""id"": 12, ""name"": ""Adventure""}, {""id"": 14, ""...","[{""id"": 270, ""name"": ""ocean""}, {""id"": 726, ""na...","[{""cast_id"": 4, ""character"": ""Captain Jack Spa...","[{""credit_id"": ""52fe4232c3a36847f800b579"", ""de..."


# 5- NLP Preprocessing & Metadata Feature Engineering :

- This is where: raw movie metadata becomes machine-understandable semantic information.

# 5.1- understanding Json like metadata :

- we have some columns containing data in a form : a string representation of structured JSON data. 

- we aregoing to extrat the mainiful data from it 

# 5.2- Genre Extraction : 

In [54]:
# step 1- Creating the Extraction function :
def convert(text):
    L = []

    for i in ast.literal_eval(text):
        L.append(i['name'])
              
    return L

In [56]:
# step 2- extracting the genre:
movies['genres'] = movies['genres'].apply(convert)

In [60]:
movies['genres'].head()

0    [Action, Adventure, Fantasy, Science Fiction]
1                     [Adventure, Fantasy, Action]
2                       [Action, Adventure, Crime]
3                 [Action, Crime, Drama, Thriller]
4             [Action, Adventure, Science Fiction]
Name: genres, dtype: object

# 5.3- Keywords Extraction :

In [61]:
# using the same convert function to extract keywords:
movies['keywords'] = movies['keywords'].apply(convert)

In [62]:
movies['keywords'].head()

0    [culture clash, future, space war, space colon...
1    [ocean, drug abuse, exotic island, east india ...
2    [spy, based on novel, secret agent, sequel, mi...
3    [dc comics, crime fighter, terrorist, secret i...
4    [based on novel, mars, medallion, space travel...
Name: keywords, dtype: object

# 5.4- cast Extraction feature engineering:

- We do NOT keep ALL actors, Because: huge noise, memory explosion, weak relevance
- Real systems often keep: top-billed actors only
- for this we will only take actors with order 1 or 0 .

In [65]:
# step 1- creating the cast extraction function 
def cast_extract(text):
    L = []

    for i in ast.literal_eval(text):
        if(i['order'] < 3):
            L.append(i['name'])

    return L

In [67]:
movies['cast'] = movies['cast'].apply(cast_extract)

In [68]:
movies['cast'].head()

0    [Sam Worthington, Zoe Saldana, Sigourney Weaver]
1       [Johnny Depp, Orlando Bloom, Keira Knightley]
2        [Daniel Craig, Christoph Waltz, Léa Seydoux]
3        [Christian Bale, Michael Caine, Gary Oldman]
4      [Taylor Kitsch, Lynn Collins, Samantha Morton]
Name: cast, dtype: object

# 5.5- Director Extraction From crew:

In [70]:
# creating the director extraction function:
def director_extrcat(text):
    L = []

    for i in ast.literal_eval(text):
        if(i['job'] == 'Director'):
            L.append(i['name'])
            break

    return L

In [72]:
movies['crew'] = movies['crew'].apply(director_extrcat)

In [73]:
movies['crew'].head()

0        [James Cameron]
1       [Gore Verbinski]
2           [Sam Mendes]
3    [Christopher Nolan]
4       [Andrew Stanton]
Name: crew, dtype: object

# 5.6- Overview text processing :

- overview is the richest semantic feature in the whole dataset.
- We want a semantic content-based recommender, So we should: KEEP semantic richness and NOT over-clean.

In [74]:
movies['overview'][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

# step 1- Remove ponctuation: 

In [80]:
def no_punc(text):
    mess = ''
    
    mess = ''.join([char for char in text if char not in string.punctuation ])

    return mess

In [84]:
movies['overview'] = movies['overview'].apply(no_punc)

In [85]:
movies['overview'].head(2)

0    In the 22nd century a paraplegic Marine is dis...
1    Captain Barbossa long believed to be dead has ...
Name: overview, dtype: object

In [86]:
movies['overview'][0]

'In the 22nd century a paraplegic Marine is dispatched to the moon Pandora on a unique mission but becomes torn between following orders and protecting an alien civilization'

- we removed the punctuations fro the overview text. they don't have semantic meaning.

# step 2- Tokinisation :
- for this w going to use a simple split() function is enough here.

In [87]:
# spliting on space:
movies['overview'] = movies['overview'].apply(lambda x : x.split())

In [88]:
movies['overview'].head(2)

0    [In, the, 22nd, century, a, paraplegic, Marine...
1    [Captain, Barbossa, long, believed, to, be, de...
Name: overview, dtype: object

- now we have overview as tokinized word representation.

# 5.7- Removing space from multiword tokens :

In [89]:
# Remove space from  multiwords in genre:
movies['genres'] = movies['genres'].apply(lambda x : [i.replace(" ", "") for i in x])

In [91]:
movies['genres'].head(2)

0    [Action, Adventure, Fantasy, ScienceFiction]
1                    [Adventure, Fantasy, Action]
Name: genres, dtype: object

In [92]:
# Same thng for keywords :
movies['keywords'] = movies['keywords'].apply(lambda x : [i.replace(" ", "") for i in x])

In [93]:
# same for cast:
movies['cast'] = movies['cast'].apply(lambda x : [i.replace(" ", "") for i in x])

In [94]:
# for crew :
movies['crew'] = movies['crew'].apply(lambda x : [i.replace(" ", "") for i in x])

- now we have words like james conner will be jamesconer and will be treated like one semantic entity.

# 5.8- Create Unified Semantic Tags :
- THIS is the HEART of the recommendation engine.

- Here We combine ALL semantic information into One unified semantic representation called: tags

In [95]:
# creating the tags column :
movies['tags'] = (
    movies['overview'] +
    movies['genres'] +
    movies['keywords'] +
    movies['cast'] +
    movies['crew']
)

In [98]:
movies['tags'].head(2)

0    [In, the, 22nd, century, a, paraplegic, Marine...
1    [Captain, Barbossa, long, believed, to, be, de...
Name: tags, dtype: object

- This is:semantic feature fusion

# 5.9- create Final NLP Dataset:

- since everything important is in the tags feature we make the new movies data set nto this :

In [99]:
new_movies = movies[['movie_id', 'title', 'tags']]

In [100]:
new_movies.head(2)

,movie_id,title,tags
0,19995,Avatar,"[In, the, 22nd, century, a, paraplegic, Marine..."
1,285,Pirates of the Caribbean: At World's End,"[Captain, Barbossa, long, believed, to, be, de..."


# 5.10- Convert tags into text :

- Currently tags are lists, and  Vectorizers need TEXT, so must convert to text.

In [101]:
new_movies['tags'] = new_movies['tags'].apply(lambda x : " ".join(x))

In [103]:
new_movies['tags'][0]

'In the 22nd century a paraplegic Marine is dispatched to the moon Pandora on a unique mission but becomes torn between following orders and protecting an alien civilization Action Adventure Fantasy ScienceFiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d SamWorthington ZoeSaldana SigourneyWeaver JamesCameron'

# 5.11- Lowercasing :
- without lowercasing Machine sees different tokens. (Action != action)

In [104]:
new_movies['tags'] = new_movies['tags'].apply(lambda x : x.lower())

In [105]:
new_movies['tags'][0]

'in the 22nd century a paraplegic marine is dispatched to the moon pandora on a unique mission but becomes torn between following orders and protecting an alien civilization action adventure fantasy sciencefiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d samworthington zoesaldana sigourneyweaver jamescameron'

# 5.12- Stemming :
Without stemming Machine sees:(love, loving, loved) as DIFFERENT words. And this is bad semantic understanding.

In [106]:
# creating the stemmer :
ps = PorterStemmer()

In [107]:
# creating the stemming function :
def stemming(text):
    y = []

    for i in text.split():
        y.append(ps.stem(i))

    return " ".join(y)

In [108]:
# applying the stemming function to tags:
new_movies['tags'] = new_movies['tags'].apply(stemming)

In [109]:
# result:
new_movies['tags'][0]

'in the 22nd centuri a parapleg marin is dispatch to the moon pandora on a uniqu mission but becom torn between follow order and protect an alien civil action adventur fantasi sciencefict cultureclash futur spacewar spacecoloni societi spacetravel futurist romanc space alien tribe alienplanet cgi marin soldier battl loveaffair antiwar powerrel mindandsoul 3d samworthington zoesaldana sigourneyweav jamescameron'

# FINAL RESULT: 

Now each movie has: a clean normalized semantic representation

READY for:

- vectorization
- embeddings
- similarity learning
- recommendations

In [111]:
new_movies.head()

,movie_id,title,tags
0,19995,Avatar,in the 22nd centuri a parapleg marin is dispat...
1,285,Pirates of the Caribbean: At World's End,captain barbossa long believ to be dead ha com...
2,206647,Spectre,a cryptic messag from bond’ past send him on a...
3,49026,The Dark Knight Rises,follow the death of district attorney harvey d...
4,49529,John Carter,john carter is a warweari former militari capt...


# 6- Text Vectorization and Similarity Modeling :

Now we will transform semantic text into mathematical vectors so machines can:

- compare movies
- measure similarity
- retrieve recommendations

# in this first version we will use the Bag-of-words method :

- bag of words It represents text using: word frequency, ignoring grammar, order and deep context.
- but for recommendation systems, semantic overlap is often enough.

# 6.1- Using CountVevtorizer: 

In [112]:
movies_cv = CountVectorizer(max_features=5000, stop_words='english')

- Keeps only top 5000 most important words. Without limit: vocabulary explode, memory usage increases, noisy rare words appear.
- also Earlier we kept StopWrods for semantic preservation, but now: CountVectorizer can remove them more intelligently.

In [113]:
# create the vectors:
vectors = movies_cv.fit_transform(new_movies['tags']).toarray()

In [116]:
vectors

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], shape=(4806, 5000))

In [117]:
# checkng the vectors shape:
vectors.shape

(4806, 5000)

- meaning 4806 movies, and 5000 word vocabulary

# 6.2- Similarity Modeling:
- Movies are similar if: their vectors point in similar directions
- We measure this mathematically.
- there a lot method to calculate the distance, but we are going to use the cosine similarity (best for NLP ) because orientation matters more than magnitude.

In [118]:
# Create the smiliraty matrix : using the Cosine similarity method : 
similarity = cosine_similarity(vectors)

In [119]:
similarity

array([[1.        , 0.08238526, 0.08492078, ..., 0.06419407, 0.02360961,
        0.        ],
       [0.08238526, 1.        , 0.06063391, ..., 0.02291746, 0.        ,
        0.02711631],
       [0.08492078, 0.06063391, 1.        , ..., 0.04724556, 0.        ,
        0.        ],
       ...,
       [0.06419407, 0.02291746, 0.04724556, ..., 1.        , 0.03940552,
        0.02112886],
       [0.02360961, 0.        , 0.        , ..., 0.03940552, 1.        ,
        0.06993786],
       [0.        , 0.02711631, 0.        , ..., 0.02112886, 0.06993786,
        1.        ]], shape=(4806, 4806))

In [120]:
similarity.shape

(4806, 4806)

- Every movie compared to every other movie. so we have a matrix with all movies in row with all movies in columns , and the value are the similarit between them (value btween 0 and 1)

# 6.3- Creating the Recomender system function :

In [136]:
def recommend(movie):

    #getting the movie index:
    movie_index = new_movies[new_movies['title'] == movie].index[0]

    #getting the similarity score for this movie index:
    distances = similarity[movie_index]

    # adding movies inces with enumerated the sorting them Descending on the similarity score using python list:
    movie_list = sorted(
        list(enumerate(distances)),
        reverse=True,
        key=lambda x:x[1]
    )[1:6]

    # after removing it self staring from 1 in the list, we loop to get their top 5 recomendetion titles.
    for i in movie_list:
        print(new_movies.iloc[i[0]].title)

In [137]:
# testing the recomender function :
recommend('Avatar')

Aliens
Titan A.E.
Independence Day
Predators
Battle: Los Angeles


- The Bag of Words model produced highly intuitive and genre-consistent recommendations. Most suggested movies strongly match Avatar through explicit sci-fi and alien-war related keywords.
- This version performs well because Avatar contains many distinctive sci-fi terms that also appear directly in these movies.
However, the recommendations are still based mainly on: surface-level word matching

# 7- TF-IDF BASED RECOMMENDATION SYSTEM : 

- Bag-Of-Words asks : 'How MANY times does a word appear?'.
- TF-IDF asks: 'How IMPORTANT is this word in this document?'.

# 7.1- TF-IDF Vectorizer:

In [138]:
# create the TF-IDF vectorizer:
tfidf_vectorizer = TfidfVectorizer(max_features=5000, stop_words='english')

In [139]:
# fiting and transforming :
tfidf_vectors = tfidf_vectorizer.fit_transform(new_movies['tags']).toarray()

In [141]:
tfidf_vectors.shape

(4806, 5000)

In [144]:
# get learned words from vectorizer :
tfidf_vectorizer.get_feature_names_out()

array(['007', '10', '100', ..., 'zone', 'zoo', 'zooeydeschanel'],
      shape=(5000,), dtype=object)

In [145]:
# see tf-idf vectors:
tfidf_vectors

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], shape=(4806, 5000))

In [148]:
# see tfidf vector on first movie:
tfidf_vectors[0]

array([0., 0., 0., ..., 0., 0., 0.], shape=(5000,))

# 7.2- Similarity Modeling:
- we also going to use Cosine Similarity:

In [149]:
tfidf_similarity = cosine_similarity(tfidf_vectors)

In [150]:
tfidf_similarity

array([[1.        , 0.02181427, 0.03036229, ..., 0.04487327, 0.00544162,
        0.        ],
       [0.02181427, 1.        , 0.01313174, ..., 0.01730405, 0.        ,
        0.00638714],
       [0.03036229, 0.01313174, 1.        , ..., 0.02326123, 0.        ,
        0.        ],
       ...,
       [0.04487327, 0.01730405, 0.02326123, ..., 1.        , 0.01529229,
        0.01890721],
       [0.00544162, 0.        , 0.        , ..., 0.01529229, 1.        ,
        0.0113172 ],
       [0.        , 0.00638714, 0.        , ..., 0.01890721, 0.0113172 ,
        1.        ]], shape=(4806, 4806))

In [151]:
tfidf_similarity.shape

(4806, 4806)

- this is movie-to-movie similarity matrix.

# 7.3- Building the TFIDF Recomender function :

In [152]:
def recommend_tfidf(movie):

    movie_index = new_movies[new_movies['title'] == movie].index[0]

    distances = tfidf_similarity[movie_index]

    movies_list = sorted(
        list(enumerate(distances)),
        reverse=True,
        key=lambda x: x[1]
    )[1:6]

    for i in movies_list:
        print(new_movies.iloc[i[0]].title)

In [153]:
# testing the TFIDF Recomender:
recommend_tfidf('Avatar')

Aliens
Battle: Los Angeles
Star Trek Into Darkness
Apollo 18
Meet Dave


- The TF-IDF system produced slightly more refined recommendations compared to Bag of Words. Because TF-IDF reduces the importance of extremely common words, the model focused more on distinctive and informative sci-fi terms.

- This improved the recommendation quality by reducing noisy generic words.

Positive observations:
- Star Trek Into Darkness is a stronger semantic sci-fi match than some Bag of Words outputs
- less dependence on overly common terms
-  more importance given to unique contextual vocabulary

However, some recommendations reveal TF-IDF limitations:
- Meet Dave appears less semantically aligned with Avatar despite sharing some space/alien-related vocabulary.

TF-IDF still lacks deep contextual understanding

# 8- WORD2VEC MOVIE RECOMMENDER SYSTEM (Semantic Embedding-Based NLP Recommender)

- istead of counting words manually, the model LEARNS meaning automatically.

# 8.1- Tokinazation for Word2Vec:
- opposite to the TF-IDF, Word2Vec does not accepted: raw text strings
- It expects: list of tokenized sentences

In [157]:
# we tokinaze our tags features and create teh tokens feature:
new_movies['tokens'] = new_movies['tags'].apply(word_tokenize)

In [159]:
#checking for tikens of each movies:
new_movies[['movie_id', 'tokens']].head()

,movie_id,tokens
0,19995,"[in, the, 22nd, centuri, a, parapleg, marin, i..."
1,285,"[captain, barbossa, long, believ, to, be, dead..."
2,206647,"[a, cryptic, messag, from, bond, ’, past, send..."
3,49026,"[follow, the, death, of, district, attorney, h..."
4,49529,"[john, carter, is, a, warweari, former, milita..."


# 8.2- Prepare the Traaining Corpus:

In [161]:
# before: 
new_movies['tokens'].head()

0    [in, the, 22nd, centuri, a, parapleg, marin, i...
1    [captain, barbossa, long, believ, to, be, dead...
2    [a, cryptic, messag, from, bond, ’, past, send...
3    [follow, the, death, of, district, attorney, h...
4    [john, carter, is, a, warweari, former, milita...
Name: tokens, dtype: object

In [162]:
# after make them into list:
sentences = new_movies['tokens'].tolist()

In [169]:
#sentences

- sentences becomes a list of movies token lists, and this is our training corpus.

# 8.3- Training the Word2Vec Model :

In [168]:
w2v_model = Word2Vec(
    sentences=sentences,
    vector_size=100,
    window=5,
    min_count=2,
    workers=4
)

We used: 
- our sentences as the curpos to train the model .
- each word becomes 100-dementional semantic vector (dense vectors).
- context window size of 5 , to capture semantic meaning of 5 word on each side .
- we told the model to ignore rare words appearing fewer than 2 times.
- we use 4 parallel cpu to train faster.

In [172]:
# cheking the vocabulary size:
len(w2v_model.wv.index_to_key)

15079

- the model learned 15079 words from the movies sentences corpus used inside the model.

In [173]:
# lets undettand each word how it represented in a dense vecotr:
w2v_model.wv['alien']

array([-6.8507707e-01, -1.5192652e-02,  6.7022282e-01, -1.3655517e-01,
       -3.4134331e-01, -1.0963799e+00,  3.1416830e-01,  1.2451955e+00,
        1.2097286e-01, -1.0115017e+00,  2.4645466e-01, -8.0925208e-01,
       -9.7708210e-02,  4.9347067e-01, -4.9253132e-02, -2.9313895e-01,
        4.6021613e-01, -6.5133798e-01,  1.3832833e-01, -9.3995816e-01,
        3.2229391e-01,  3.7793919e-01,  4.7584057e-01, -3.9733833e-01,
       -1.0759243e-01,  5.8995831e-01, -2.7203330e-01,  1.7675716e-02,
       -6.2958390e-01,  1.1764648e-01,  6.7953223e-01,  4.5756161e-01,
       -2.2532072e-02, -2.9568300e-01, -4.6608707e-01,  7.0289439e-01,
        7.1288127e-01, -1.0242141e+00, -4.2879102e-01, -5.4354852e-01,
        4.9122743e-02, -6.5752530e-01,  3.8264713e-01, -4.0894818e-01,
        6.1364329e-01, -7.9133689e-02, -6.2044144e-01, -1.5238421e-01,
        3.3600554e-01,  9.1106452e-02,  7.5123549e-01,  1.2537748e-01,
       -9.7909756e-02, -4.8207700e-01, -8.4779561e-01,  1.4722031e-01,
      

- this is the 100-dementional sematic dense vector for the word 'alien' for exemple.This is semantic embedding.

In [174]:
# let's also check for similar words to alien :
w2v_model.wv.most_similar('alien')

[('chase', 0.9989703297615051),
 ('disast', 0.9988490343093872),
 ('technolog', 0.9983248710632324),
 ('supernatur', 0.9982511401176453),
 ('boat', 0.9979874491691589),
 ('pennsylvania', 0.9978480339050293),
 ('cult', 0.9977141618728638),
 ('isol', 0.9976862668991089),
 ('expedit', 0.9976117610931396),
 ('mercenari', 0.997590184211731)]

- this proves that semantic learning happend, and this works because : words used in similar contexts become nearby vectors.
This is distributional semantics.

# 8.4- Movie Embedding :

- Word2Vec gives words embbedings, but we need movie embedding to create vectors for each movie so we can compre it to each other not just words.
- to do this we are going to average all word vectors inside each movie. to create the movie vector.

In [175]:
# creating the movie embedding function :
def document_vector(doc):

    # chekcing if the movie words has its embedding vector in our model leanred vocab.
    words = [word for word in doc if word in w2v_model.wv]

    # if all words dont exist we create the 100-dementional vector with all zeros.
    if len(words) == 0:
        return np.zeros(100)

    # returning the movie vector as average of all words vector inside it.
    return np.mean(w2v_model.wv[words], axis=0)

In [179]:
# now using the function to ceating the movies embeddings:
movie_vectors = np.array(new_movies['tokens'].apply(document_vector).tolist())

In [180]:
# checking the shape of it :
movie_vectors.shape

(4806, 100)

- matrix with all movies in row each row a single semantic dense vector of 100 demention 

# 8.5- Modeling the Similarity with Cosine similarity :

In [181]:
w2v_similarity = cosine_similarity(movie_vectors)

- same logic because the same recommender pipline , only the vector representation that change:

In [182]:
w2v_similarity.shape

(4806, 4806)

# 8.6- Creating the recommander function :

In [183]:
def recommend_w2v(movie):

    movie_index = new_movies[new_movies['title'] == movie].index[0]

    distances = w2v_similarity[movie_index]

    movies_list = sorted(
        list(enumerate(distances)),
        reverse=True,
        key=lambda x:x[1]
    )[1:6]

    for i in movies_list:
        print(new_movies.iloc[i[0]].title)

In [184]:
# testing the Word2Vec recomender :
recommend_w2v('Avatar')

Beowulf
Southland Tales
X-Men Origins: Wolverine
Reno 911!: Miami
13 Hours: The Secret Soldiers of Benghazi


- This result is extremely important because it reveals both the power and limitations of basic Word2Vec document averaging.

The recommendations became much less keyword-driven and more influenced by learned semantic vector relationships. This proves the model moved beyond simple word overlap.

However, the results also became noticeably unstable and less relevant to Avatar.

But major issues appeared:
- several recommendations are weak thematic matches
- Reno 911!: Miami is clearly unrelated to Avatar
- averaging word embeddings caused semantic dilution
- movie-level embeddings lost important contextual structure

This happened because:

simple averaging of word vectors destroys sentence structure and contextual meaning

The model understands word relationships, but your document representation method is still primitive.

The next systems you build will progressively solve the weaknesses observed in Word2Vec by introducing:


# 9- FastText NLP Recommender System :

- the Word2Vec algorithme has some limitation and most important ones are Out-Of-Vocabulary error and morfology blinsspots as it treat each word as one vectors and new words or typos won't be recognized.
- here it comes fasttext with the solution which is instead of using a word embedding, FastText uses Subword embedding, and the calculate each word buy combaining (summing) the vectors of its constituents. 